crates/sui/src/client_ptb/ast.rs

```diff
@@ -0,0 +1,353 @@
+pub const U256: &str = "u256";
+
+// Keywords
+pub const ADDRESS: &str = "address";
+pub const BOOL: &str = "bool";
+pub const VECTOR: &str = "vector";
+pub const SOME: &str = "some";
+pub const NONE: &str = "none";
+pub const GAS: &str = "gas";
+
+pub const KEYWORDS: &[&str] = &[
+    ADDRESS, BOOL, VECTOR, SOME, NONE, GAS, U8, U16, U32, U64, U128, U256,
+];
+
+pub fn is_keyword(s: &str) -> bool {
+    KEYWORDS.contains(&s)
+}
+
+pub fn all_keywords() -> String {
+    KEYWORDS[..KEYWORDS.len() - 1]
+        .iter()
+        .map(|x| format!("'{}'", x))
+        .collect::<Vec<_>>()
+        .join(", ")
+        + &format!(", or '{}'", KEYWORDS[KEYWORDS.len() - 1])
+}
+
+/// A PTB Program consisting of a list of commands and a flag indicating if the preview
+/// warn-shadows command was present.
+#[derive(Debug, Clone)]
+pub struct Program {
+    pub commands: Vec<Spanned<ParsedPTBCommand>>,
+    // Held outside of metadata since this is used by the PTB builder
+    pub warn_shadows_set: bool,
+}
+
+/// The `ProgramMetadata` struct holds metadata about a PTB program, such as whether the preview
+/// flag was set, json output was set, etc.
+#[derive(Debug, Clone)]
+pub struct ProgramMetadata {
+    pub preview_set: bool,
+    pub summary_set: bool,
+    pub gas_object_id: Option<Spanned<ObjectID>>,
+    pub json_set: bool,
+    pub gas_budget: Spanned<u64>,
+}
+
+/// A parsed module access consisting of the address, module name, and function name.
+#[derive(Debug, Clone)]
+pub struct ModuleAccess {
+    pub address: Spanned<ParsedAddress>,
+    pub module_name: Spanned<Identifier>,
+    pub function_name: Spanned<Identifier>,
+}
+
+/// A parsed PTB command consisting of the command and the parsed arguments to the command.
+#[derive(Debug, Clone)]
+pub enum ParsedPTBCommand {
+    TransferObjects(Spanned<Argument>, Spanned<Vec<Spanned<Argument>>>),
+    SplitCoins(Spanned<Argument>, Spanned<Vec<Spanned<Argument>>>),
+    MergeCoins(Spanned<Argument>, Spanned<Vec<Spanned<Argument>>>),
+    MakeMoveVec(Spanned<ParsedType>, Spanned<Vec<Spanned<Argument>>>),
+    MoveCall(
+        Spanned<ModuleAccess>,
+        Option<Spanned<Vec<ParsedType>>>,
+        Vec<Spanned<Argument>>,
+    ),
+    Assign(Spanned<String>, Option<Spanned<Argument>>),
+    Publish(Spanned<String>),
+    Upgrade(Spanned<String>, Spanned<Argument>),
+    WarnShadows,
+    Preview,
+}
+
+/// An enum representing the parsed arguments of a PTB command.
+#[derive(Debug, Clone)]
+pub enum Argument {
+    Bool(bool),
+    U8(u8),
+    U16(u16),
+    U32(u32),
+    U64(u64),
+    U128(u128),
+    U256(move_core_types::u256::U256),
+    Gas,
+    Identifier(String),
+    VariableAccess(Spanned<String>, Vec<Spanned<String>>),
+    Address(NumericalAddress),
+    String(String),
+    Vector(Vec<Spanned<Argument>>),
+    Option(Spanned<Option<Box<Argument>>>),
+}
+
+impl Argument {
+    /// Resolve an `Argument` into a `MoveValue` if possible. Errors if the `Argument` is not
```

Comment:
Either the argument will resolve into a `MoveValue` or it will error if a corresponding `MoveValue` cannot be found (e.g., for an identifier, variable access etc).